macaw-aarch32/src/Data/Macaw/ARM/Arch.hs

```diff
@@ -137,25 +133,60 @@ instance TF.FunctorF ARMTermStmt where
 instance TF.TraversableF ARMTermStmt where
   traverseF go tstmt =
     case tstmt of
-      ARMSyscall imm -> pure (ARMSyscall imm)
-      ThumbSyscall imm -> pure (ThumbSyscall imm)
       ReturnIf cond -> ReturnIf <$> go cond
       ReturnIfNot cond -> ReturnIfNot <$> go cond
 
 rewriteTermStmt :: ARMTermStmt (MC.Value ARM.AArch32 src) -> Rewriter ARM.AArch32 s src tgt (ARMTermStmt (MC.Value ARM.AArch32 tgt))
 rewriteTermStmt s =
     case s of
-      ARMSyscall imm -> pure $ ARMSyscall imm
-      ThumbSyscall imm -> pure (ThumbSyscall imm)
       ReturnIf cond -> ReturnIf <$> rewriteValue cond
       ReturnIfNot cond -> ReturnIfNot <$> rewriteValue cond
 
 -- ----------------------------------------------------------------------
 -- ARM functions.  These may return a value, and may depend on the
--- current state of the heap and the set of registeres defined so far
+-- current state of the heap and the set of registers defined so far
 -- and the result type, but should not affect the processor state.
 
 data ARMPrimFn (f :: MT.Type -> Type) tp where
+  -- | Issue a system call
+  --
+  -- The intent is that the user provides a mapping from system call numbers to
+  -- handlers in macaw-aarch32-symbolic, enabling the translation to crucible to
+  -- replace this operation with a lookup + call to a function handle.  By
+  -- capturing all of the necessary registers as inputs and outputs here,
+  -- uniform treatment is possible.  See the x86 version for a more detailed
+  -- account of the translation strategy.
+  --
+  -- The 'WI.W' is the immediate operand embedded in the opcode
+  --
+  -- The rest of the arguments are all of the registers that participate in the
+  -- syscall protocol (at least on Linux). Arguments are passed in r0-r6, while
+  -- the syscall number is in r7.
+  --
+  -- The system call can return up to two values (in r0 and r1)
+  ARMSyscall :: WI.W 24
+             -> f (MT.BVType 32) -- r0
+             -> f (MT.BVType 32) -- r1
+             -> f (MT.BVType 32) -- r2
+             -> f (MT.BVType 32) -- r3
+             -> f (MT.BVType 32) -- r4
+             -> f (MT.BVType 32) -- r5
+             -> f (MT.BVType 32) -- r6
+             -> f (MT.BVType 32) -- r7 (syscall number)
+             -> ARMPrimFn f (MT.TupleType [MT.BVType 32, MT.BVType 32])
```

Comment:
Since you have comments stating which registers each of the arguments correspond to, you might as well add comments for the return registers as well. Similarly for `ThumbSyscall`'s return registers.